service/inject/maven-plugin/src/main/java/io/helidon/service/inject/maven/plugin/BindingGenerator.java

```diff
@@ -0,0 +1,584 @@
+                    .addContent(".class, ")
+                    .addContent(providerImpl)
+                    .addContent("::new, ")
+                    .addContent(String.valueOf(sl.weight()))
+                    .addContent(")");
+        } else {
+            // the usual singleton instance
+            return it -> it.addContent(serviceInfo.descriptorType().fqName())
+                    .addContent(".INSTANCE");
+        }
+    }
+
+    private static Lookup toLookup(TypeName typeName) {
+        return Lookup.builder()
+                .serviceType(typeName)
+                .build();
+    }
+
+    /**
+     * Determines if the service is valid to receive injections.
+     *
+     * @param sp the service provider
+     * @return true if the service provider can receive injection
+     */
+    private static boolean isQualifiedInjectionTarget(InjectServiceInfo sp) {
+        Set<ResolvedType> contractsImplemented = sp.contracts();
+        List<Ip> dependencies = sp.dependencies();
+
+        if (contractsImplemented.contains(ResolvedType.create(INJECT_BINDING))) {
+            return false;
+        }
+        boolean hasDependencies = !dependencies.isEmpty();
+        boolean hasContract = !contractsImplemented.isEmpty();
+
+        return hasContract || hasDependencies;
+    }
+
+    private TypeInfo createAppTypeInfo(TypeName typeName) {
+        return TypeInfo.builder()
+                .kind(ElementKind.CLASS)
+                .typeName(typeName)
+                // to trigger generation of descriptor
+                .addAnnotation(Annotation.create(SERVICE_ANNOTATION_PROVIDER))
+                .addInterfaceTypeInfo(TypeInfo.builder()
+                                              .kind(ElementKind.INTERFACE)
+                                              .typeName(INJECT_BINDING)
+                                              .addAnnotation(Annotation.create(SERVICE_ANNOTATION_CONTRACT))
+                                              .build())
+                .build();
+    }
+
+    private void handleError(CodegenException ce) {
+        if (failOnError) {
+            throw ce;
+        } else {
+            ctx.logger().log(ce.toEvent(System.Logger.Level.WARNING));
+        }
+    }
+
+    private InjectionPlan injectionPlan(WrappedServices services,
+                                        InjectServiceInfo self,
+                                        Ip dependency) {
+        /*
+         very similar code is used in ServiceManager.planForIp
+         make sure this is kept in sync!
+         */
+        Lookup dependencyTo = Lookup.create(dependency);
+        Set<Qualifier> qualifiers = dependencyTo.qualifiers();
+        if (self.contracts().containsAll(dependencyTo.contracts()) && self.qualifiers().equals(qualifiers)) {
+            // criteria must have a single contract for each injection point
+            // if this service implements the contracts actually required, we must look for services with lower weight
```

Comment:
This is related to injection points - if an injection point requires a contract that the current service implements, we must look for lower weight, as otherwise the current service would be selected to satisfy the injection point of itself.

Reply:
I have updated the comment with the above statement.